File: src/presentation/routes/gentube.py
from apiflask import APIBlueprint
from flask import jsonify, request, send_file

from src.domain.services import gentube_animation_service
from src.infrastructure.ai_providers import gentube_bridge
from src.presentation.schemas.gentube import GentubeRunPromptsInSchema
from src.utils.logger import get_logger

logger = get_logger(__name__)

gentube_bp = APIBlueprint("gentube", __name__, url_prefix="/api/gentube")


# ─────────────────────────────────────────────────────────────────
# Bridge de deteccion de sesion (background.js hace chrome.cookies.getAll en
# gentube.app y postea aca -- ver gentube_bridge.py y el comentario del alarm
# "hb" en extensions/flow_extension/background.js).
# ─────────────────────────────────────────────────────────────────


@gentube_bp.post("/bridge-register")
def bridge_register():
    data = request.get_json(silent=True) or {}
    probe = gentube_bridge.set_session_from_cookie(data.get("cookie", ""))
    return jsonify({"ok": bool(probe.get("ok"))})


@gentube_bp.get("/status")
def status():
    return jsonify(gentube_animation_service.get_status())


@gentube_bp.get("/check-login")
def check_login():
    return jsonify(profiles=gentube_animation_service.check_login())


@gentube_bp.post("/login")
def login():
    data = request.get_json(force=True, silent=True) or {}
    account_id = int(data.get("profile", data.get("account_id", 0)))
    try:
        result = gentube_animation_service.start_login(account_id, data.get("cookie", ""))
        return jsonify(result)
    except ValueError as exc:
        return jsonify(error=str(exc)), 400


@gentube_bp.post("/run-prompts")
@gentube_bp.input(GentubeRunPromptsInSchema)
def run_prompts(json_data):
    try:
        result = gentube_animation_service.start_run(
            json_data["prompts"],
            json_data["slots"],
            json_data["repeat"],
            json_data["output_dir"],
            json_data["ratio"],
            json_data["quality"],
            browser_mode=json_data.get("browser_mode", "chromium"),
        )
        return jsonify(result)
    except (ValueError, RuntimeError) as exc:
        status_code = 409 if isinstance(exc, RuntimeError) else 400
        return jsonify({"error": str(exc)}), status_code


@gentube_bp.post("/stop")
def stop():
    gentube_animation_service.stop()
    return jsonify({"ok": True})


@gentube_bp.post("/reset")
def reset():
    gentube_animation_service.reset()
    return jsonify({"ok": True})


@gentube_bp.get("/images")
def images():
    return jsonify(gentube_animation_service.list_images(request.args.get("dir")))


@gentube_bp.get("/image/<path:name>")
def image(name):
    path = gentube_animation_service.get_image_path(name, request.args.get("dir"))
    if not path:
        return jsonify({"error": "not found"}), 404
    return send_file(path)


@gentube_bp.post("/clear-images")
def clear_images():
    data = request.get_json(silent=True) or {}
    gentube_animation_service.clear_images(data.get("output_dir"))
    return jsonify({"ok": True})

File: frontend/src/api/gentube.ts
import { api } from "./client";

export interface GentubeProfile {
  name?: string;
  logged_in?: boolean;
}

export interface GentubeCheckLoginResult {
  profiles?: GentubeProfile[];
}

export interface GentubeStatus {
  running?: boolean;
  log?: string[];
  total?: number;
  processed?: number;
  images?: number;
  rate?: string | number;
}

export interface GentubeImagesResult {
  images?: string[];
}

export type GentubeBrowserMode = "chrome" | "chromium";

export interface GentubeRunPromptsParams {
  prompts: string[];
  slots?: number;
  repeat?: number;
  output_dir?: string;
  ratio?: string;
  quality?: string;
  browser_mode?: GentubeBrowserMode;
}

export function gentubeStatus() {
  return api.get<GentubeStatus>("/gentube/status").then((r) => r.data);
}

export function gentubeCheckLogin() {
  return api.get<GentubeCheckLoginResult>("/gentube/check-login").then((r) => r.data);
}

export function gentubeLogin(
  { profile = 0, cookie = "" }: { profile?: number; cookie?: string } = {},
) {
  return api
    .post<{ ok: boolean }>("/gentube/login", { profile, cookie })
    .then((r) => r.data);
}

export function gentubeRunPrompts({
  prompts,
  slots = 1,
  repeat = 1,
  output_dir = "",
  ratio = "1:1",
  quality = "standard",
  browser_mode = "chromium",
}: GentubeRunPromptsParams) {
  return api
    .post<{ ok: boolean }>("/gentube/run-prompts", {
      prompts,
      slots,
      repeat,
      output_dir,
      ratio,
      quality,
      browser_mode,
    })
    .then((r) => r.data);
}

export function gentubeStop() {
  return api.post<{ ok: boolean }>("/gentube/stop").then((r) => r.data);
}

export function gentubeReset() {
  return api.post<{ ok: boolean }>("/gentube/reset").then((r) => r.data);
}

export function gentubeImages(dir?: string) {
  return api
    .get<string[] | GentubeImagesResult>("/gentube/images", { params: dir ? { dir } : {} })
    .then((r) => r.data);
}

export function gentubeImageUrl(name: string, dir?: string) {
  const params = new URLSearchParams();
  if (dir) params.set("dir", dir);
  const qs = params.toString();
  return `/api/gentube/image/${encodeURIComponent(name)}${qs ? `?${qs}` : ""}`;
}

export function gentubeClearImages(output_dir?: string) {
  return api
    .post<{ ok: boolean }>("/gentube/clear-images", output_dir ? { output_dir } : {})
    .then((r) => r.data);
}
